repair/repair.cc

```diff
@@ -1600,3 +1600,92 @@ future<> bootstrap_with_repair(seastar::sharded<database>& db, locator::token_me
+            auto& strat = ks.get_replication_strategy();
+            // First get all ranges the leaving node is responsible for
+            dht::token_range_vector ranges = strat.get_ranges(leaving_node);
+            rlogger.info("{}: started with keyspace={}, leaving_node={}, nr_ranges={}", op, keyspace_name, leaving_node, ranges.size());
+            std::unordered_map<dht::token_range, std::vector<inet_address>> current_replica_endpoints;
+            // Find (for each range) all nodes that store replicas for these ranges as well
+            for (auto& r : ranges) {
+                auto end_token = r.end() ? r.end()->value() : dht::maximum_token();
+                auto eps = strat.calculate_natural_endpoints(end_token, tm);
+                current_replica_endpoints.emplace(r, std::move(eps));
+            }
+            auto temp = tm.clone_after_all_left();
+            // leaving_node might or might not be 'leaving'. If it was not leaving (that is, removenode
+            // command was used), it is still present in temp and must be removed.
+            if (temp.is_member(leaving_node)) {
+                temp.remove_endpoint(leaving_node);
+            }
+            std::unordered_map<dht::token_range, repair_neighbors> range_sources;
+            for (auto it = ranges.begin(); it != ranges.end();) {
+                auto& r = *it;
+                if (need_preempt()) {
+                    seastar::thread::yield();
+                }
+                auto end_token = r.end() ? r.end()->value() : dht::maximum_token();
+                const std::vector<inet_address> new_eps = ks.get_replication_strategy().calculate_natural_endpoints(end_token, temp);
+                const std::vector<inet_address>& current_eps = current_replica_endpoints[r];
+                std::unordered_set<inet_address> neighbors_set(new_eps.begin(), new_eps.end());
+                bool skip_this_range = false;
+                // For decommission operation, the decommission node will
+                // repair all the ranges the leaving node is responsible for.
+                if (!is_decommission) {
+                    // For removenode operation, the new owner of the range
+                    // will repair the data. Nodes that are not the new owner
+                    // of a range will ignore the range. There can be only one
+                    // new owner for each of the ranges.
+                    auto new_owner = neighbors_set;
+                    for (const auto& node : current_eps) {
+                        new_owner.erase(node);
+                    }
+                    if (new_owner.size() != 1) {
+                        rlogger.warn("{}: keyspace={}, range={}, current_replica_endpoints={}, new_replica_endpoints={}",
+                            op, keyspace_name, r, current_eps, new_eps);
+                        throw std::runtime_error("Can not find new owner node!");
```

Comment:
It is actually a new owner node. The new owner node == new replica nodes - old replica nodes. e.g., new replica nodes = n1,n2,n3, old replica nodes = n1,n2,n4, new owner node = n3. Please note, when a node is removed, there must be a new node to replace the removed node, so there will always be a new node in the new replica nodes list.

Reply:
By "owner" I refer a node that would be a primary replica or the node that would own the corresponding token range that contains the token in question. 

If old replicas set was n1, n2, n3 and the new one is n1, n2, n4 - the new owner node is not necessarily going to be n4 - it may be n1 or n2. 

A very simple example is a Simple Strategy with RF=3. When you remove a node the right boundary of the corresponding range is removed and the node that was the first non-primary replica before becomes the primary replica of corresponding tokens.

Reply:
I know what you mean. But the "new owner node" does not necessarily mean it is the new node that holds the primary replica.  It only means it is the new owner of one the replica.

Reply:
I added more comments to make it clear.

Reply:
you could simply call the variable "new_replica" and it would not require additional comments... ;)